tests/disaster-recovery/regional-dr/test_failover.py

```diff
@@ -0,0 +1,93 @@
+import logging
+import pytest
+
+from time import sleep
+
+from ocs_ci.framework import config
+from ocs_ci.framework.testlib import rdr_test
+from ocs_ci.helpers import dr_helpers
+from ocs_ci.ocs.exceptions import CommandFailed
+from ocs_ci.ocs.node import wait_for_nodes_status, get_node_objs
+from ocs_ci.ocs.ocp import wait_for_cluster_connectivity
+
+logger = logging.getLogger(__name__)
+
+
+@rdr_test
+class TestFailover:
+    @pytest.fixture()
+    def setup(self, request, nodes, rdr_workload):
+        dr_helpers.set_primary_cluster_context(rdr_workload.workload_namespace)
+        primary_cluster_index = config.cur_index
+        all_nodes = get_node_objs()
+
+        def finalizer():
+            config.switch_ctx(primary_cluster_index)
+            try:
+                wait_for_cluster_connectivity(tries=3)
+            except CommandFailed:
+                nodes.start_nodes(all_nodes)
+                wait_for_nodes_status([node.name for node in all_nodes])
+
+        request.addfinalizer(finalizer)
+
+        return all_nodes, rdr_workload
+
+    @pytest.mark.parametrize(
+        argnames=["primary_cluster_down"],
+        argvalues=[
+            pytest.param(False),
+            pytest.param(True),
+        ],
+    )
+    def test_failover(self, nodes, setup, primary_cluster_down):
+        """
+        Test Failover action
+
+        """
+        all_nodes, rdr_workload = setup
+
+        scheduling_interval = dr_helpers.get_scheduling_interval(
+            rdr_workload.workload_namespace
+        )
+        logger.info(f"Waiting for {2 * scheduling_interval} minutes for IOs to run")
+        sleep(2 * scheduling_interval * 60)
+
+        if primary_cluster_down:
+            dr_helpers.set_primary_cluster_context(rdr_workload.workload_namespace)
+            nodes.stop_nodes(all_nodes)
+
+        secondary_cluster_name = dr_helpers.get_secondary_cluster_name(
+            rdr_workload.workload_namespace
+        )
+        dr_helpers.failover(secondary_cluster_name, rdr_workload.workload_namespace)
```

Comment:
@sidhant-agrawal , we need to come up with context switch consistency semantics. For ex: Here in the line # 57 you are setting context to primary, when we call `dr_helpers.failover()` context will be switch to ACM cluster by the helper function. For someone looking into this testcase flow, they might assume that after line # 63 we are still in primary cluster context which is wrong because helper function has changed the context to ACM. Either helper should should restore the context before return OR we need framework wide mechanism to control context switches. Let me know your thoughts ! @clacroix12 @petr-balogh @prsurve Need your inputs on this.

Reply:
I agree.
+1 to framework wide mechanism to control context switches if that's possible.
Or otherwise we would need to update all helper methods which are changing context to restore it before return.

Reply:
All relevant helper methods in dr_helpers.py are now restoring the context.